File: src/java/eu/sig/training/ch03/withmapandtypes/FlagFactory.java
package eu.sig.training.ch03.withmapandtypes;

import static eu.sig.training.ch03.Nationality.BELGIAN;
import static eu.sig.training.ch03.Nationality.DUTCH;
import static eu.sig.training.ch03.Nationality.FRENCH;
import static eu.sig.training.ch03.Nationality.GERMAN;
import static eu.sig.training.ch03.Nationality.ITALIAN;

import java.awt.Color;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import eu.sig.training.ch03.Nationality;

public class FlagFactory {

    // tag::getFlag[]
    private static final Map<Nationality, Flag> FLAGS =
        new HashMap<Nationality, Flag>();

    static {
        FLAGS.put(DUTCH, new DutchFlag());
        FLAGS.put(GERMAN, new GermanFlag());
        FLAGS.put(BELGIAN, new BelgianFlag());
        FLAGS.put(FRENCH, new FrenchFlag());
        FLAGS.put(ITALIAN, new ItalianFlag());
    }

    public List<Color> getFlagColors(Nationality nationality) {
        Flag flag = FLAGS.get(nationality);
        flag = flag != null ? flag : new DefaultFlag();
        return flag.getColors();
    }
    // end::getFlag[]

}

File: src/csharp/eu/sig/training/ch03/withmapandtypes/FlagFactory.cs
using System.Collections.Generic;
using System.Drawing;

namespace eu.sig.training.ch03.withmapandtypes
{

    public class FlagFactory
    {

        // tag::getFlag[]
        private static readonly Dictionary<Nationality, IFlag> FLAGS =
            new Dictionary<Nationality, IFlag>();

        static FlagFactory()
        {
            FLAGS[Nationality.DUTCH] = new DutchFlag();
            FLAGS[Nationality.GERMAN] = new GermanFlag();
            FLAGS[Nationality.BELGIAN] = new BelgianFlag();
            FLAGS[Nationality.FRENCH] = new FrenchFlag();
            FLAGS[Nationality.ITALIAN] = new ItalianFlag();
        }

        public IList<Color> GetFlagColors(Nationality nationality)
        {
            IFlag flag = FLAGS[nationality];
            flag = flag ?? new DefaultFlag();
            return flag.Colors;
        }
        // end::getFlag[]

    }

}
